torch/csrc/autograd/python_variable_indexing.cpp

```diff
@@ -369,51 +330,72 @@ int THPVariable_setitem(PyObject* self, PyObject* index, PyObject* py_value) {
 
   auto& self_ = reinterpret_cast<THPVariable*>(self)->cdata;
   OptionalDeviceGuard device_guard(device_of(self_));
+  at::Device self_device = self_.device();
   Variable value;
   // TODO: This qint special case looks very suspicious...
   if (isQIntType(self_.scalar_type())) {
-    value = valueToTensor(device(kCPU).dtype(kFloat), py_value);
+    value = valueToTensor(device(kCPU).dtype(kFloat), py_value, at::Device(kCPU));
   } else {
-    value = valueToTensor(self_.options(), py_value);
+    value = valueToTensor(self_.options(), py_value, self_device);
   }
+  bool is_tracing = torch::jit::tracer::isTracing();
+
+  auto handle_simple_type = [&](at::indexing::TensorIndex&& tensor_index, Variable&& value) {
```

Comment:
You can move capture `value` in the lambda, avoiding the need for the argument. SEe https://stackoverflow.com/questions/8640393/move-capture-in-lambda

Reply:
But more to the point, the move of `value` into the lambda is pointless as you do not actually move it into `set_item`. Did you mean to `std::move(value)`?

Reply:
Checking on the type of `set_item`, I see that it takes `const Tensor&`. So all of the faffing about with the rvalue reference is unnecessary, just capture it as a normal reference and be done with it.

Reply:
It seems that the lambda `handle_simple_type` actually adds overhead to the indexing path, so I removed it in favor of direct function call to `at::indexing::get_item / set_item`